src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -960,8 +963,13 @@ async Task<IEnumerable<IAccount>> ITokenCacheInternal.GetAccountsAsync(Authentic
                 allEnvironmentsInCache,
                 requestParameters.RequestContext).ConfigureAwait(false);
 
-            rtCacheItems = rtCacheItems.Where(rt => instanceMetadata.Aliases.ContainsOrdinalIgnoreCase(rt.Environment)).ToList();
-            accountCacheItems = accountCacheItems.Where(acc => instanceMetadata.Aliases.ContainsOrdinalIgnoreCase(acc.Environment)).ToList();
+            // If the client application is instance aware then we skip the filter with environment
+            // since the authority in request is different from the authority used to get the token
+            if (!requestParameters.AppConfig.MultiCloudSupport)
+            {
+                rtCacheItems = rtCacheItems.Where(rt => instanceMetadata.Aliases.ContainsOrdinalIgnoreCase(rt.Environment)).ToList();
+                accountCacheItems = accountCacheItems.Where(acc => instanceMetadata.Aliases.ContainsOrdinalIgnoreCase(acc.Environment)).ToList();
```

Comment:
Should we call `ToList()` to make the code more consistent?

Reply:
Maybe we can ToList in the foreach loops below this code, instead of here? And the Verbose logging on 974 can be moved after foreach also, since it calls Count.

Although I'm not sure if calls to .Count and .ToList reevaluate the expression tree each time (probably since expression can have reference to some external delegate code that can change result each time).

Reply:
@pmaytak I tried to remove ToList() from here and it will give a compilation error `rtCacheItems` is an object of type `IReadOnlyList` and `rtCacheItems.Where` returns an object of type `IEnumerable` so the `ToList` is used to convert it to `List` object. It will not impact any of the flow below.